KAFKA-3452: Support session windows

dguy opened the pull request:
Add support for SessionWindows based on design detailed in https://cwiki.apache.org/confluence/display/KAFKA/KIP-94+Session+Windows.
This includes refactoring of the RocksDBWindowStore such that functionality common with the RocksDBSessionStore isn't duplicated.

dguy committed:
session windows

dguy commented:
@enothereska @mjsax @guozhangwang.

This follows the design as proposed in the KIP. 
Some key things:
 I've extracted all of the `Segment` functionality out of `RocksDBWindowStore` and into `Segments`. I've introduced a new class `SegmentedBytesStore` that both the `RocksDBWindowStore` and `RocksDBSessionStore` delegate to. They are wrappers on top of the `SegmentedBytesStore` that know how to deal with the specifics to those stores, i.e., iterators, key layout etc.
The new class `SessionWindows` doesn't derive from `Windows` as it doesn't make sense to pass a `SessionWindows` instance to the non-sesssion-window aggregate methods on `KGroupedStream`

Thanks

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -98,6 +99,42 @@
+    /**
+     * Combine values of this stream by key into {@link SessionWindows}
+     * The resulting {@link KTable} will be materialized in a local state
+     * store with the given store name. Also a changelog topic named "${applicationId}-${storeName}-changelog"
```
StoreName is not defined in this API.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -179,6 +216,59 @@
+    /**
+     * Aggregate values of this stream by key into {@link SessionWindows}.
+     * The resulting {@link KTable} will be materialized in a local state
+     * store with the given store name. Also a changelog topic named "${applicationId}-${storeName}-changelog"
```
Same here with storename.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -231,4 +321,34 @@
+    /**
+     * Count number of records of this stream by key into {@link SessionWindows}.
+     * The resulting {@link KTable} will be materialized in a local state
+     * store with the given store name. Also a changelog topic named "${applicationId}-${storeName}-changelog"
```
Same here with storename.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,95 @@
+ * +-----------+-------------+------------+
+ *
+ * The previous 2 sessions would be merged into a single session with start time 10 and end time 20.
+ * The aggregate value for this session would be the result of aggregating all 4 values.
```
Nice explanation!

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionKeyBinaryConverter.java:
```diff
@@ -0,0 +1,64 @@
+import org.apache.kafka.streams.kstream.Windowed;
+
+import java.nio.ByteBuffer;
+
```
Consider adding a line or two on what this class is meant for.

mjsax replied:
This is a general thing we lack for many internal classes... It's hard to get started if it's not clear what a class does and even more important *why*.

The comment added is actually quite good!

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TupleForwarder.java:
```diff
@@ -25,21 +25,24 @@
 class TupleForwarder<K, V> {
```
Can we add a line to describe what this class is for?

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/LRUCacheEntry.java:
```diff
@@ -50,7 +50,7 @@
                 8 + // offset
                 4 + // partition
-                topic.length();
+                (topic == null ? 0 : topic.length());
```
Good catch!

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:
```diff
@@ -106,16 +106,23 @@ synchronized void flush() {
             }
             entries.add(new ThreadCache.DirtyEntry(key, node.entry.value, node.entry));
             node.entry.markClean();
+            if (node.entry.value == null) {
```
So if value==null we automatically delete?

dguy replied:
Yes - automatically delete from the cache.

enothereska reviewed (commented):
Initial pass on just high level APIs and high-level implementation only.

dguy committed:
added comments as per feedback from eno

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionMerger.java:
```diff
@@ -0,0 +1,37 @@
+     * @param aggTwo    the second aggregate
+     * @return          the new aggregate value
+     */
+    V apply(K aggKey, V aggOne, V aggTwo);
```
Do we need the key?

dguy replied:
It is consistent with `Aggregator` - also, why not? Maybe a developer wants to do something with the key.

mjsax replied:
It's a matter of taste I guess. For example `ValueJoiner` has no key argument either (there was a discussion about adding the key though...). The API is nor really consistent right now. Should we try to make it consisted adding key parameter everywhere?
\cc @enothereska @miguno @guozhangwang

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,95 @@
+
+
+/**
+ * A session based window specification uses for aggregating events into sessions.
```
nit: uses -> used

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,95 @@
+ * then a new session will be created.
+ *
+ * For example, If we have a session gap of 5 and the following data arrives:
+ * +--------------------------------------+
```
nit: this will not render nicely as compiled JavaDoc. Can you add some markup?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,95 @@
+     * @return a new SessionWindows with the provided inactivity gap
+     * and default maintain duration
+     */
+    public static SessionWindows inactivityGap(final long gapMs) {
```
What about `with(final long inactivityGap)` instead of `inactivityGap(final long gapMs)`? This would be closer to `TimeWindows#of(long size)` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -188,6 +192,170 @@ public Long apply(K aggKey, V value, Long aggregate) {
+        final SessionMerger<K, V> sessionMerger = new SessionMerger<K, V>() {
+            @Override
+            public V apply(final K aggKey, final V aggOne, final V aggTwo) {
+                return aggregator.apply(aggKey, aggTwo, aggOne);
```
We have to document that the aggregator is used a SessionMerger.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    public long size() {
+        return gapMs;
```
The "gap" is not the size... For me, a `size()` method does not make any sense for session windows. Only a instance of a session window does have a current size.

dguy replied:
Yep  - i agree and it can probably be removed now.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -188,6 +192,170 @@ public Long apply(K aggKey, V value, Long aggregate) {
+
+
+    private <W extends Window, T> StateStoreSupplier windowedStore(final Serde<T> aggValSerde,
+                                                                   final Windows<W> windows,
```
Why not make this `SessionWindows` instead of `Window` and use `InternalSessionWindows` (or actually an anonymous class) only within this method.

Here it get's clear why you need `InternalSessionWindows` in the first place -- it's confusion above.

dguy replied:
Sorry, but i don't follow. This method is used for other Window stores. Are you suggesting a different method?

dguy replied:
I take it back - it isn't used anywhere else. The method actually moved to `AbstractStream`, bad merge.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -0,0 +1,161 @@
+
+    private TimeWindow mergeTimeWindow(final TimeWindow one, final TimeWindow two) {
+        final long start = one.start() < two.start() ? one.start() : two.start();
+        final long end = one.end() < two.end() ? two.end() : one.end();
```
This spins my head... can we change to ?
`final long end = one.end() > two.end() ? one.end() : two.end();`
or
`final long end = two.end() > one.end()? two.end() : one.end();`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TupleForwarder.java:
```diff
@@ -21,25 +21,34 @@
+                   final boolean sendOldValues) {
         this.cached = store instanceof CachedStateStore;
         this.context = context;
+        this.sendOldValues = sendOldValues;
```
Why make this a member?

dguy replied:
Because the value of `sendOldValues` is always fixed at construction time, i.e, when `Processor.init(..)` is called. So pass it in here rather than with every method invocation

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionMerger.java:
```diff
@@ -0,0 +1,37 @@
+ * @param <K>   key type
+ * @param <T>   aggregate value type
+ */
+public interface SessionMerger<K, V> {
```
Add annotation "instable"

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,95 @@
+ * The previous 2 sessions would be merged into a single session with start time 10 and end time 20.
+ * The aggregate value for this session would be the result of aggregating all 4 values.
+ */
+public class SessionWindows {
```
Add annotation "instable"

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -0,0 +1,161 @@
+            TimeWindow mergedWindow = newTimeWindow;
+            T agg = initializer.apply();
+
+            try (final KeyValueIterator<Windowed<K>, T> iterator = store.findSessionsToMerge(key, timestamp - windows.gap() + 1,
```
JavaDoc says "Fetch any session aggregates with the matching sessionId and the sessions end is >= earliestEndTime and the sessions start is <= latestStartTime"

So why `+1` and `-1` ?

dguy replied:
The javadoc is correct.
It is to do with the way i've interpreted what to do with the gap ( which is possibly confusing).
i.e, if the gap is 5 and 5 seconds has elapsed it should be a new Session. So given a timestamp of 15 i only want to find sessions where earliestEndTime >= 11 and the latestStartTime <= 19

dguy replied:
I've changed this to not do the `+1` and `-1` as it probably makes more sense.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java:
```diff
@@ -0,0 +1,47 @@
+public interface SessionStore<K, AGG> extends StateStore, ReadOnlySessionStore<K, AGG> {
+
+    /**
+     * Fetch any session aggregates with the matching sessionId and the sessions end is >= earliestEndTime and the sessions
```
`sessionId` -> `key`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java:
```diff
@@ -0,0 +1,47 @@
+     * Fetch any session aggregates with the matching sessionId and the sessions end is >= earliestEndTime and the sessions
+     * start is <= latestStartTime
+     */
+    KeyValueIterator<Windowed<K>, AGG> findSessionsToMerge(K sessionId, long earliestSessionEndTime, long latestSessionStartTime);
```
`sessionId` -> `key`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -188,6 +192,170 @@ public Long apply(K aggKey, V value, Long aggregate) {
+                                                                   final boolean sessionWindows,
+                                                                   final String storeName) {
+        return storeFactory(keySerde, aggValSerde, storeName)
+                .windowed(windows.size(), windows.maintainMs(), windows.segments, false, sessionWindows)
```
Why does the the session gap (=windows.size()) define the size of the underlying window store?

dguy replied:
It doesn't really - the param is ignored when it is creating a session store. However, i went back and forth with using the `windowed` function or adding a new `sessionWindows` function to the `PersistentKeyValueFactory` - i couldn't really decide which approach i liked best, but i'm thinking i should add the function. Thoughts?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -62,7 +62,7 @@
  * @see org.apache.kafka.streams.state.Stores#create(String)
  */
-public class RocksDBStore<K, V> implements KeyValueStore<K, V> {
+class RocksDBStore<K, V> implements KeyValueStore<K, V> {
```
Why not allow developers to use it?

dguy replied:
It is in the `internals` package and i believe in making internal code as private as possible. If they want to use the `RocksDBStore` they can already create one via the various suppliers etc.

All that said, though, this could break anyone that has sub-classed this. So i'll change it back.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:
```diff
@@ -416,6 +429,194 @@ public int compare(final KeyValue<String, Long> o1, final KeyValue<String, Long>
+                        new Properties()),
+                t1);
+
+        final long t2 = t1 + (sessionGap / 2);
```
`final long t2 = t1 + sessionGap;`
to test the corner case

dguy replied:
nah - that is what i wanted. There are further tests in `KStreamSessionWindowAggregateProcessorTest`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:
```diff
@@ -416,6 +429,194 @@ public int compare(final KeyValue<String, Long> o1, final KeyValue<String, Long>
+
+    }
+
 
```
Could you add a test, that does merge two session into a single one, because a late record arrives (ie late record is in between both session) ?

dguy replied:
Is already covered in  `KStreamSessionWindowAggregateProcessorTest`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationIntegrationTest.java:
```diff
@@ -416,6 +429,194 @@ public int compare(final KeyValue<String, Long> o1, final KeyValue<String, Long>
+                        StringSerializer.class,
+                        new Properties()),
+                t2);
+        final long t3 = t1 + sessionGap;
```
`final long t3 = t1 + sessionGap + 1`; -- otherwise I would assume that the sessions get merged.

dguy replied:
if the `sessionGap` has passed then a new Session is created.

mjsax commented on streams/src/test/java/org/apache/kafka/test/MockProcessorContext.java:
```diff
@@ -157,7 +161,7 @@ public StateStore getStateStore(String name) {
     @Override
     public void schedule(long interval) {
-        throw new UnsupportedOperationException("schedule() not supported.");
+//        throw new UnsupportedOperationException("schedule() not supported.");
```
???

dguy replied:
Good question!

dguy committed:
address comments

dguy committed:
rename getter on SessionWindows

dguy committed:
change the session gap semantics

dguy committed:
add sessionWindows method to PersistentKeyValueFactory

dguy committed:
rename SessionMerger -> Merger

dguy committed:
merge with trunk

dguy committed:
merge trunk

dguy committed:
fix checkstyle

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -360,6 +424,107 @@
+                                  final String storeName);
+
+    /**
+     * Combine values of this stream by the grouped key into {@link SessionWindows}.
```
Do we need repeated Javadoc everywhere? Not sure what's best practices here. Just there is a lot of repetition and perhaps one aggregate can have the Javadoc and others can point to it?

dguy replied:
I know - it is consistent with the rest of the class.

mjsax replied:
Agree that it is a problem, but if you use an IDE and hoover over a class, it should display it -- and not just a link to another method... I would be awesome if JavaDoc would allow to factor common stuff out for overloaded methods. But I don't know a good way to do it either -- there is only c&p... :(

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindows.java:
```diff
@@ -0,0 +1,105 @@
+     *
+     * @return  itself
+     */
+    public SessionWindows until(long durationMs) {
```
Should duration be `final`?

dguy replied:
This one yes.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java:
```diff
@@ -0,0 +1,47 @@
+     * Fetch any session aggregates with the matching key and the sessions end is >= earliestEndTime and the sessions
+     * start is <= latestStartTime
+     */
+    KeyValueIterator<Windowed<K>, AGG> findSessionsToMerge(K key, long earliestSessionEndTime, long latestSessionStartTime);
```
Should parameters be `final` for consistency with your other code?

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -179,6 +180,69 @@
+     *         where each table contains records with unmodified keys and values
+     *         that represent the latest (rolling) count (i.e., number of records) for each key within that window
+     */
+    KTable<Windowed<K>, Long> count(final SessionWindows sessionWindows, final String storeName);
```
This probably doesn't belong to this PR, but I wonder why we need count at all, it's just an aggregate.

mjsax replied:
I guess, it's THE aggregate everyone wants. Syntactic sugar.

dguy committed:
comments

dguy committed:
merge trunk and comments

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -0,0 +1,161 @@
+                while (iterator.hasNext()) {
+                    final KeyValue<Windowed<K>, T> next = iterator.next();
+                    merged.add(next);
+                    agg = sessionMerger.apply(key, agg, next.value);
```
This would have probably better been part of the KIP discussion, but there will be some aggregates, like `Average` for which this kind of merging won't work because they will need more than just the previous aggregate values. How do we catch such cases and warn the user?

dguy replied:
This is the same for non SessionWindows based aggregates. I.e, if you aggregate an int into an Average you cant just apply the new int to the currently aggregated Average.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -0,0 +1,161 @@
+            if (!mergedWindow.equals(newTimeWindow)) {
+                for (final KeyValue<Windowed<K>, T> session : merged) {
+                    store.remove(session.key);
+                    tupleForwarder.maybeForward(session.key, null, session.value);
```
Don't we want to first invalidate previous values, then send new values, i.e., flip the order of this forwarding with the previous one? I can't immediately find the Javadoc where we tell the user the order.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java:
```diff
@@ -0,0 +1,196 @@
+        };
+    }
+
+    static class SessionStoreIterator<K, AGG> implements KeyValueIterator<Windowed<K>, AGG> {
```
Why is this part of RocksDbSessionStore? Isn't it useful for stores other than RocksDb-based?

dguy replied:
It doesn't strictly need to be part of `RocksDbSessionStore`, but it is only used here as of now. If it is needed elsewhere later, then we should move it later.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -0,0 +1,229 @@
+class CachingSessionStore<K, AGG>  implements SessionStore<K, AGG>, CachedStateStore<Windowed<K>, AGG> {
+
+    private final SegmentedBytesStore bytesStore;
+    private final RocksDBSessionStore.SessionKeySchema keySchema;
```
If we use a different store, not RocksDb, will this variable still be good?

dguy replied:
Yes - it applies to a `SegmentedBytesStore` - which is not `RocksDB` specific. The class could be moved out of RocksDBSessionStore, but I don't really see the need to at the moment.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -0,0 +1,229 @@
+            public void apply(final List<ThreadCache.DirtyEntry> entries) {
+                for (ThreadCache.DirtyEntry entry : entries) {
+                    final Bytes binaryKey = entry.key();
+                    final RecordContext current = context.recordContext();
```
In previous stores, like `CachingKeyValueStore`, you put the code below in a separate method called `putAndMaybeForward` that I thought was a neat way to break the code into chunks.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    public void close() {
+        bytesStore.close();
```
Do we want to flush and close the cache too?

dguy replied:
Yeah. The closing of the cache was dependent on another PR that has now been merged. `flush` doesn't matter so much as we always `flush` and then `close`, but i'll add it anyway

dguy committed:
clear cache on close. address some feedback

enothereska commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -188,6 +191,137 @@ public Long apply(K aggKey, V value, Long aggregate) {
+        Objects.requireNonNull(sessionMerger, "sessionMerger can't be null");
+        Objects.requireNonNull(storeSupplier, "storeSupplier can't be null");
+
+        return (KTable<Windowed<K>, T>) doAggregate(
```
Is the casting needed?

dguy replied:
It is consistent with what is already there. Also, i can't think of another way of doing it without duplicating `doAggregate`, which i'd rather not do

enothereska commented:
@dguy I had a look over:
- APIs
- main algo for merging sessions
- session store and IQ read only session store
- kgroupedstreamimpl
- basic skimming over new tests (they look great!).

Apart from the above comments I don't have anything else. Thanks.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java:
```diff
@@ -179,6 +180,69 @@
+     *         where each table contains records with unmodified keys and values
+     *         that represent the latest (rolling) count (i.e., number of records) for each key within that window
+     */
+    KTable<Windowed<K>, Long> count(final SessionWindows sessionWindows, final String storeName);
```
It's a bit unclear to me as `it doesn't make sense to pass a SessionWindows instance to the non-sesssion-window aggregate methods on KGroupedStream`, could you elaborate it a bit more why we would really want overloaded functions here? Is it because we already did so in the general `aggregate` functions because of the additional `sessionMerger` parameter and we want to be consistent elsewhere?

dguy replied:
I'm not sure what you mean here? `SessionWindows` doesn't extend `Windows`. This is a deliberate decision as what you have written above is true:

> it doesn't make sense to pass a SessionWindows instance to the non-sesssion-window aggregate methods on KGroupedStream

So if we want to provide a `count` for `SessionWindows` we need an overloaded method.

guozhangwang replied:
Yeah I think I understand this now, I was thinking ahead of time and was wondering that if we added Merger into other aggregations as well with a different underlying impl, could we save this overloaded method for `aggregate` and `count` etc. But that might be to far away to discuss now, and even if we do that, we would have to do it in a backward incompatible way anyways.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -19,21 +19,24 @@
 import org.apache.kafka.streams.kstream.Reducer;
+import org.apache.kafka.streams.kstream.Merger;
+import org.apache.kafka.streams.kstream.SessionWindows;
+import org.apache.kafka.streams.kstream.Window;
```
nit: could you order the imports?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:
```diff
@@ -188,6 +191,137 @@ public Long apply(K aggKey, V value, Long aggregate) {
+                                                      aggregator,
+                                                      sessionMerger);
+
+        return (KTable<Windowed<K>, Long>) doAggregate(
```
nit: could we call `aggregate` with the passed appliers here to be consistent with other callers?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -0,0 +1,160 @@
+            if (!mergedWindow.equals(newTimeWindow)) {
+                for (final KeyValue<Windowed<K>, T> session : merged) {
+                    store.remove(session.key);
+                    tupleForwarder.maybeForward(session.key, null, session.value);
```
Why we want to forward the values for merged and newly added windows (i.e. here and line 113)? Should we only forward when we "close" them?

dguy replied:
We discussed that at length during the design phase. We decided that we'd stick with the same forwarding as we do with everything else as it is consistent. 
It is actually listed as a rejected alternative in the KIP

guozhangwang replied:
I remember that now. Thanks.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionKeyBinaryConverter.java:
```diff
@@ -0,0 +1,70 @@
+ * {@link Windowed} key used in the store.
+ * @param <K>
+ */
+public class SessionKeyBinaryConverter<K> {
```
Could we just inline these functionalities in `SessionWindowedSerializer` and `SessionWindowedDeserializer`? 

Also @bbejeck is working on https://issues.apache.org/jira/browse/KAFKA-4468 which is semi-related to this class as well, in which case the end timestamp is not stored but will only be calculated for other typed windows. I think in the future we can still combining them into base class / extended classes when we add merger for other windowed aggregations and hence we can let SessionWindows to extend from Windows. Just FYI.

dguy replied:
The end timestamp can't be calculated for `SessionWindows` as it is not of a fixed length. We need to store both start and end in this case

guozhangwang replied:
Yes I know. As I said the difference is that for timed windows, the window size need to be remembered in the `Windows` object and the underlying RocksDB (or any other store engine) only need to serialize the start timestamp in to persistent bytes, where in session windows there is no window size to remember and the underlying RocksDB needs to serialize both start and end timestamp. Personally I think these two different types of implementation can just be captured in separate serializes (e.g. WindowedSerializer and SessionWindowSerializer, or more generally named `BoundedWindowedSerializer` and `PartialBoundedWindowedSerializer`).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -29,9 +29,9 @@
-import org.apache.kafka.streams.state.internals.RocksDBWindowStoreSupplier;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.apache.kafka.streams.state.internals.WindowStoreSupplier;
```
nit: group this with other o.a.k. imports?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -105,19 +106,28 @@ public StateStoreSupplier build() {
                                     }
 
                                     @Override
+                                    public PersistentKeyValueFactory<K, V> sessionWindows(final long retentionPeriod) {
```
nit: rename to "sessionWindowed" and the above function to `timeWindowed`? Ditto in line 407.

dguy replied:
i think `sessionWindowed` is ok, but renaming `windowed` to `timeWindowed` is a breaking API change. Can we do that without a KIP?

guozhangwang replied:
Good point. Maybe just renaming this to `sessionWindowed` then for now.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -0,0 +1,160 @@
+                if (!iter.hasNext()) {
+                    return null;
+                }
+                return iter.next().value;
```
Validate this iterator only contain one element?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -0,0 +1,243 @@
+
+    public void remove(final Windowed<K> sessionKey) {
+        validateStoreOpen();
+        put(sessionKey, null);
```
Any reason we do not use cache.delete and underlying.delete like modified elsewhere?

dguy replied:
`delete` on `CachingKeyValueStore` is not used anywhere except for tests - it probably should be removed. If we delete from a store we always put key:null.
I think there was another reason, too. But i can't remember what it was. It has been quite some time since this was written

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java:
```diff
@@ -0,0 +1,196 @@
+        return findSessionsToMerge(key, 0, Long.MAX_VALUE);
+    }
+
+    static class SessionKeySchema implements RocksDBSegmentedBytesStore.KeySchema {
```
Is this specific to RocksDB, since it is used in the general CachingSessionState as well?

dguy replied:
No it is not specific to RocksDB. The `implements` should be `SegmentedBytesStore.KeySchema` - i'll move it out of this class

guozhangwang reviewed (commented):
Thanks @dguy for the big patch. Made a first pass over the PR.

I have a general thought that may be addressed separately from this PR: our current state store implementation along with caching and dedupping has getting very very complicated with layered design of iterators / caches, etc. Many of the classes are quite similar and it has been very hard (at least for myself) to get the hierarchy of classes. 

I have filed KAFKA-4587	to rethink about caching and dedupping, but more generally speaking I think it is possible to:

1. Have the wrappers of metered -> logged -> cached -> underlying along with the similar iterator wrappers as well for all types of state stores, and similarly for queryable state hierarchies as well, hence condensing the class files. 

2. Still let SessionWindows to extend from Windows eventually with mergers that are applied to non-session-windowed aggregation as well. In this case, we do not need to treat session window stores as a completely different stack but similar to window stores with different key-serde manners.

3. Make caching configurable per store, and capture the forwarding logic in the caching layer with evict listeners and get rid of the hacky tuple forwarders.

dguy committed:
address comments

dguy committed:
address comments

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreSupplier.java:
```diff
@@ -0,0 +1,21 @@
+ */
+package org.apache.kafka.streams.state.internals;
+
+public interface WindowStoreSupplier {
```
Is it cleaner to let `WindowStoreSupplier` extend `StateStoreSupplier`, and let `RocksDBSessionStoreSupplier` and `RocksDBWindowStoreSupplier` implement `WindowStoreSupplier`. In this way we can add a `PersistentWindowFactory` in `Stores` whose `build()` then returns `WindowStoreSupplier`?

dguy replied:
I'm pretty sure adding a `PersistentWindowFactory` would be a breaking change. `Stores` is a public API and to make this work we'd need to change `PersistentKeyValueFactory.windowed` to return a `PersistentWindowFactory` - this could break existing usages. Would this need another KIP?

guozhangwang replied:
I agree, and I think `PersistentKeyValueFactory<K, V> windowed` is the wrong API since people can still call sth. like:

```
...windowed(...).sessionWindowed(...)
```

and it is quite confusing to use and I'd rather fix it now than later. Although it changes public APIs I believe most people will be using it directly followed by a `build()` call, so I think it is OK to do it along with this KIP as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java:
```diff
@@ -0,0 +1,47 @@
+     * Fetch any session aggregates with the matching key and the sessions end is >= earliestEndTime and the sessions
+     * start is <= latestStartTime
+     */
+    KeyValueIterator<Windowed<K>, AGG> findSessionsToMerge(final K key, long earliestSessionEndTime, final long latestSessionStartTime);
```
Maybe rename to "findSessions" as they are not necessarily for merging? Also it seems not relevant to "session aggregates"? Should the java doc be "fetch any sessions with the matching key and the end timestamp >= ... "?

dguy replied:
In the KIP it was named `findSessionsToMerge`, but i agree they are not necessarily for merging. Should i change it and update the KIP?
As for the comment, sessions, session aggregates, same thing really. But i can just change it sessions

guozhangwang replied:
Yeah let's update the KIP correspondingly with the renaming.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java:
```diff
@@ -0,0 +1,47 @@
+public interface SessionStore<K, AGG> extends StateStore, ReadOnlySessionStore<K, AGG> {
+
+    /**
+     * Fetch any session aggregates with the matching key and the sessions end is >= earliestEndTime and the sessions
```
nit: use `&le` etc to be safer with different browser.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSessionStore.java:
```diff
@@ -0,0 +1,143 @@
+        return findSessionsToMerge(key, 0, Long.MAX_VALUE);
+    }
+
+    static class SessionStoreIterator<K, AGG> implements KeyValueIterator<Windowed<K>, AGG> {
```
This can be private.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java:
```diff
@@ -0,0 +1,49 @@
+    void put(Bytes key, byte[] value);
+
+    @Override
+    boolean isOpen();
```
Why this override?

dguy replied:
accident

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentIterator.java:
```diff
@@ -0,0 +1,107 @@
+        throw new UnsupportedOperationException("remove not supported");
+    }
+
+    interface KeyExtractor<K> {
```
Are they used only for testing purposes? If yes could we add some comments explaining their usage?

dguy replied:
They are not needed. Left overs from some early work i did. Will remove

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSegmentedBytesStore.java:
```diff
@@ -0,0 +1,177 @@
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.KeyValueIterator;
+
+public class MeteredSegmentedBytesStore implements SegmentedBytesStore {
```
We can remove `public` here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSegmentedBytesStore.java:
```diff
@@ -0,0 +1,177 @@
+    private Sensor getTime;
+    private Sensor removeTime;
+
+    public MeteredSegmentedBytesStore(final SegmentedBytesStore inner, String metricScope, Time time) {
```
We can remove `public` here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSegmentedBytesStore.java:
```diff
@@ -0,0 +1,177 @@
+        private final Sensor sensor;
+        private final long startNs;
+
+        public MeteredSegmentedBytesStoreIterator(final KeyValueIterator<Bytes, byte[]> iter, Sensor sensor) {
```
We can remove `public` here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java:
```diff
@@ -0,0 +1,49 @@
+
+public interface SegmentedBytesStore extends StateStore {
+    @SuppressWarnings("unchecked")
+    KeyValueIterator<Bytes, byte[]> fetch(Bytes key, long from, long to);
```
Better adding some comments here to explain the expected logic.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    void initInternal(final InternalProcessorContext context) {
```
Can be private.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentIterator.java:
```diff
@@ -0,0 +1,107 @@
+    @SuppressWarnings("unchecked")
+    public boolean hasNext() {
+        boolean hasNext = false;
+        while ((currentIterator == null || !(hasNext = hasNextCondition.hasNext(currentIterator)) || !currentSegment.isOpen())
```
Not sure if it is correct: `currentIterator` is in type `<Bytes, byte[]>`, while in `RocksDBWindowStore` it is `HasNextCondition<Long> ITERATOR_HAS_NEXT = new HasNextCondition<Long>()` with type `<Long, ?>`. Would this work?

dguy replied:
The `HasNextCondition` interface shouldn't have any type params. It should be just `Bytes`. But, yes it does work due to erasure

guozhangwang replied:
Should we rename it then or move it into SegmentedByteStore? Please see my other comment.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/HasNextCondition.java:
```diff
@@ -16,10 +16,8 @@
-public interface PeekingWindowIterator<V> extends WindowStoreIterator<V> {
-
-    KeyValue<Long, V> peekNext();
+interface HasNextCondition<K> {
```
This interface is a bit awkward to me: in session windows it is typed `Bytes` while in timed windows it is typed `Long`: also I am not sure if typed `Long` is correct since the passed in iterator is always in `<Bytes, byte[]>`.

I'm feeling it is not worthwhile to make it an interface and we could probably just implement the specific hasNext inlined in these two classes?

dguy replied:
The interface is needed. It is used in `SegmentIterator`, for example, to determine if we should continue iterating. The type params are incorrect and i'll fix them

guozhangwang replied:
Okay, could we then rename it to sth like `SegmentedBytesStoreIteratorHasNextCondition`, or just move this interface into `SegmentedBytesStore.KeySchema`?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSessionStoreTest.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Test
+    public void shouldFindSessionsToMerge() throws Exception {
```
nit: `shouldFindSessionsInRange`?

guozhangwang commented on streams/src/test/java/org/apache/kafka/test/KeyValueIteratorStub.java:
```diff
@@ -0,0 +1,56 @@
+
+import java.util.Iterator;
+
+public class KeyValueIteratorStub<K, V> implements KeyValueIterator<K, V> {
```
nit: rename to `MockKeyValueIterator`?

dguy replied:
I prefer not to have loads of classes starting with Mock. Just makes it easier to find them in IntelliJ etc if it is the other way around

guozhangwang commented on streams/src/test/java/org/apache/kafka/test/ReadOnlySessionStoreStub.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.List;
+import java.util.Map;
+
+public class ReadOnlySessionStoreStub<K, V> implements ReadOnlySessionStore<K, V>, StateStore {
```
Ditto above.

guozhangwang commented on streams/src/test/java/org/apache/kafka/test/SegmentedBytesStoreStub.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class SegmentedBytesStoreStub implements SegmentedBytesStore {
```
Ditto above? Generally speaking we use `MockXX` in other places and just to be consistent.

dguy replied:
Yeah - i know that has been done, but i disagree with it. If you feel strongly about it i'll (begrudgingly) change them!

guozhangwang replied:
Haha :P I'm not feeling strongly about it just trying to be consistent (only paranoids survives!), but since we already have a bunch of `Stubs` there's no point I guess. Let's keep it as is.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/internals/SegmentsTest.java:
```diff
@@ -0,0 +1,193 @@
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class SegmentsTest {
```
Great coverage!

guozhangwang reviewed (commented):
@dguy I have some thoughts about minor refactoring the hierarchy a bit in this PR, let me know what do you think (other detailed comments):

1. Add a `WrappedStateStore` interface extending `StateStore` with one additional API `inner()` returning the inner most storage engine.

2. Add `LoggedKeyValueStore` / `LoggedWindowStore` / `LoggedSessionStore` and `CachingKeyValueStore` / `CachingWindowStore` / `CachingSessionStore` in addition to `MeteredKeyValueStore` and `MeteredWindowStore` and `MeteredSessionStore`, implementing `KeyValueStore` / `WindowStore` / `SessionStore` respectively and `WrappedStateStore`, where the `inner()` function recursively returns the inner-most storage engine: by doing this we can work around the restoration callback which should only call the inner most store's put (i.e. `putInternal`) function. As for iterators, we still keep them in corresponding wrapper classes.

3. Let `WindowStoreSupplier` extend StateStoreSupplier, and let `RocksDBSessionStoreSupplier` and `RocksDBWindowStoreSupplier` implement `WindowStoreSupplier`. Add a `PersistentWindowFactory` in `Stores` whose `build()` then returns `WindowStoreSupplier`, also add an `AbstractWindowStoreSupplier` that implements `WindowStoreSupplier` and extends `AbstractStoreSupplier`.

4. Let `RocksDBWindowStoreSupplier` and `RocksDBSessionStoreSupplier` extends `AbstractWindowStoreSupplier`, and their `get()` function can return multi-layer wrapped `WindowStore` and `SessionStore`.

5. `RocksDBWindowStore` and `RocksDBSessionStore` will still embed `RocksDBSegmentedBytesStore`, which does not need to have its wrappers since all the wrappers are on the parent `RocksDBWindowStore` and `RocksDBSessionStore`, and `RocksDBSegmentedBytesStore` itself is used only as the underlying storage engine.

6. Session store util classes can still be kept as separate classes like `SegmentIterator` and `Segments` (personally I feel `Segment` can be inlined in `RocksDBSegmentedBytesStore` but not strong opinion).

7. This is minor: for inlined `static class WindowStoreKeySchema` we can also make it a separate class as the `SessionKeySchema` and use it in `CachingWindowStore` as we used the other schema class in `CachingSessionStore` as well.


I know it will take longer to merge the PR, and honestly I am not sure if this factoring really works. But it does work I feel it is worthwhile to do it now than later. And we could probably clean it up more when having per-store caching mechanism.

dguy committed:
address most comments

dguy committed:
address comments

dguy commented:
@guozhangwang i've addressed your comments.
w.r.t your suggested refactoring: though i agree the hierarchy could do with some tidying up, i'm not 100% sure what you are suggesting will work and a lot of it is un-related to this PR. I think it needs some more thought and this PR has been outstanding for nearly 2 months already. Also, the changes you are suggesting are internal (i.e., not affecting public APIs), so they could be done once this PR is merged. I think we might already have a JIRA for creating Logged**Stores ? Will have to look.

dguy committed:
minor refactoring to remove some duplication

guozhangwang commented:
> I think it needs some more thought and this PR has been outstanding for nearly 2 months already.

I understand... Just afraid such tech debt will keep growing and eventually come back and bite us as unstable and hard-to-debug code. Well how about merging this PR as is and see if we can still push it as a separate PR before the release deadline?

dguy commented:
@guozhangwang - that is fine with me. I agree with the tech-debt and i think we should really focus the next release on bug fixing, tech-debt, usability etc. 
If we get this in i'm more than happy to tackle your suggestions above

guozhangwang commented:
@dguy Cool, will merge it as is. Note that I have a couple follow-up comments in the previous round, but I think it would be easier to address them as a new PR than this 5000+ LOC one.

MarcoAbi commented:
Hello,

i'm having an issue trying to use SessionWindows

This exception is thrown:

`Exception in thread "StreamThread-1" org.apache.kafka.streams.errors.StreamsException: stream-thread [StreamThread-1] Failed to rebalance
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:612)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:368)
Caused by: java.lang.IndexOutOfBoundsException
	at java.nio.Buffer.checkIndex(Buffer.java:546)
	at java.nio.HeapByteBuffer.getLong(HeapByteBuffer.java:416)
	at org.apache.kafka.streams.kstream.internals.SessionKeySerde.extractEnd(SessionKeySerde.java:117)
	at org.apache.kafka.streams.state.internals.SessionKeySchema.segmentTimestamp(SessionKeySchema.java:45)
	at org.apache.kafka.streams.state.internals.RocksDBSegmentedBytesStore.put(RocksDBSegmentedBytesStore.java:71)
	at org.apache.kafka.streams.state.internals.RocksDBSegmentedBytesStore$1.restore(RocksDBSegmentedBytesStore.java:104)
	at org.apache.kafka.streams.processor.internals.ProcessorStateManager.restoreActiveState(ProcessorStateManager.java:230)
	at org.apache.kafka.streams.processor.internals.ProcessorStateManager.register(ProcessorStateManager.java:193)
	at org.apache.kafka.streams.processor.internals.AbstractProcessorContext.register(AbstractProcessorContext.java:99)
	at org.apache.kafka.streams.state.internals.RocksDBSegmentedBytesStore.init(RocksDBSegmentedBytesStore.java:101)
	at org.apache.kafka.streams.state.internals.ChangeLoggingSegmentedBytesStore.init(ChangeLoggingSegmentedBytesStore.java:68)
	at org.apache.kafka.streams.state.internals.MeteredSegmentedBytesStore.init(MeteredSegmentedBytesStore.java:66)
	at org.apache.kafka.streams.state.internals.RocksDBSessionStore.init(RocksDBSessionStore.java:78)
	at org.apache.kafka.streams.state.internals.CachingSessionStore.init(CachingSessionStore.java:97)
	at org.apache.kafka.streams.processor.internals.AbstractTask.initializeStateStores(AbstractTask.java:86)
	at org.apache.kafka.streams.processor.internals.StreamTask.<init>(StreamTask.java:141)
	at org.apache.kafka.streams.processor.internals.StreamThread.createStreamTask(StreamThread.java:834)
	at org.apache.kafka.streams.processor.internals.StreamThread$TaskCreator.createTask(StreamThread.java:1207)
	at org.apache.kafka.streams.processor.internals.StreamThread$AbstractTaskCreator.retryWithBackoff(StreamThread.java:1180)
	at org.apache.kafka.streams.processor.internals.StreamThread.addStreamTasks(StreamThread.java:937)
	at org.apache.kafka.streams.processor.internals.StreamThread.access$500(StreamThread.java:69)
	at org.apache.kafka.streams.processor.internals.StreamThread$1.onPartitionsAssigned(StreamThread.java:236)
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:255)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:339)
	at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:303)
	at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:286)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollOnce(KafkaConsumer.java:1030)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:995)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:582)
	... 1 more`

Can you please help me with this?

Thanks.
Marco

mjsax commented:
@MarcoAbi Could you please report this at the Kafka mailing list: http://kafka.apache.org/contact

Thanks.
